src/flow_graph.py

```diff
@@ -1056,95 +1137,125 @@ def duplicate_premature_returns(nodes: List[Node]) -> List[Node]:
             assert isinstance(node.successor, ReturnNode)
             node.successor.parents.remove(node)
             index += 1
-            n = ReturnNode(node.successor.block.clone(), False, index=index)
+            n = ReturnNode(
+                node.successor.block.clone(),
+                False,
+                index=index,
+                terminal=node.successor.terminal,
+            )
             node.successor = n
             n.add_parent(node)
+            n.terminal.add_parent(n)
             extra_nodes.append(n)
 
     nodes += extra_nodes
     nodes.sort(key=lambda node: node.block.index)
-    return [n for n in nodes if n.parents or n == nodes[0]]
-
-
-def ensure_fallthrough(nodes: List[Node]) -> None:
-    """For any node which is only reachable through indirect jumps (switch
-    labels, loop edges, emit_goto edges), mark its predecessor as emit_goto to
-    ensure we continue to generate code after it."""
-    last_reachable: Set[Node] = set()
-    while True:
-        # Traverse all edges other than indirect jumps, and collect the targets.
-        reachable: Set[Node] = {nodes[0]}
-        predecessors: Dict[Node, Node] = {}
-        for i, node in enumerate(nodes):
-            fallthrough: Optional[Node]
-            if i + 1 < len(nodes):
-                fallthrough = nodes[i + 1]
-                predecessors[fallthrough] = node
+    # Filter out unreachable nodes (except the entry & terminal nodes)
```

Comment:
Yeah, tbh I think I'd lean towards removing this filter entirely? (Though I'd want to check how often the current filter is needed to make the flowgraph reducible)

For now, it's needed so that `translate_graph_from_block` in translate.py can iteratively hit every node.

Reply:
Adding TODO